Do not return nonexistent kinds in `SyntaxFacts`' methods

DoctorKrolic opened the pull request:
Fixes: https://github.com/dotnet/roslyn/issues/72300

DoctorKrolic commented on src/Compilers/CSharp/Test/Syntax/LexicalAndXml/LexicalTests.cs:
```diff
@@ -3603,6 +3603,19 @@ public void TestMaxKeywordLength()
         }
 
+        [Fact]
+        public void TestNoKind8441InKeywordKinds()
```
I didn't find any specific test file for `SyntaxFacts` class, so I put it here (e.g. test above uses `SyntaxFacts`)

DoctorKrolic replied:
Actually, found out that there is `SyntaxTests` for that. Moving the test

Youssef1313 commented on src/Compilers/CSharp/Portable/Syntax/SyntaxKindFacts.cs:
```diff
@@ -1148,7 +1148,9 @@ public static IEnumerable<SyntaxKind> GetContextualKeywordKinds()
             for (int i = (int)SyntaxKind.YieldKeyword; i <= (int)SyntaxKind.FileKeyword; i++)
             {
-                yield return (SyntaxKind)i;
+                // 8441 corresponds to a deleted kind (DataKeyword) that was previously shipped.
```
Maybe add `Debug.Assert(((SyntaxKind)i).ToString() != i.ToString())`

DoctorKrolic replied:
Simplified your suggestion to `Debug.Assert(Enum.IsDefined(typeof(SyntaxKind), (SyntaxKind)i))` (had to use version with `typeof` due to the fact, that generic overload is not present in `netstandard2.0`). Does the same things, but reads a lot easier. Also since there are multiple methods, which loops through numbers like this added such guard to all of them. Not sure, whether this will be classified as an "unrelated" change by the compiler team, but since we faced bug of this nature, I guess, it's better to add such guard rails everywhere

DoctorKrolic replied:
Adding additional asserts revealed that are a lot unexisting kinds are returned from `GetPunctuationKinds()` method

AlekseyTs commented:
> Do not return unexisting kinds in SyntaxFacts' methods

Could you please elaborate what the problem is and open an issue that describes a user-facing problem? Even though there is no member in the enum, the number is still reserved. And

DoctorKrolic commented:
> Could you please elaborate what the problem is

I was working on a fix for https://github.com/dotnet/roslyn/issues/72258. I decided to populate all C# keywords' texts upfront, so I can later return them as completions. In order to achieve that I used `SyntaxFacts.GetKeywordKinds()` method and then saved keyword text by calling `SyntaxFacts.GetText(kind)` on each kind. When it came to verifying my work I noticed that there was 1 completion with empty text. A quick debug session showed that `GetKeywordKinds` has unexisting kind `8441` as part of returned collection of kinds. This led to this PR being opened. Then Youssef suggested adding guarding assert and I added such for all methods with similar integer looping logic. Failed CI then showed that `SyntaxFacts.GetPunctuationKinds()` also returns unexisting kinds and a lot of them. After fixing it I thought that corresponding `SyntaxFacts.IsXXX` checks should also not assume anything about unexisting kinds. `SyntaxFacts.IsKeywordKind()` already specifies all contextual keywords 1 by 1, so it didn't need fixing. `SyntaxFacts.IsPunctuation()` on the other hand did a range check, so I modified it to exclude unexisting kinds there. Now we are at this state.

> open an issue that describes a user-facing problem

> Even though there is no member in the enum, the number is still reserved

It is perfectly fine to reserve number ranges for different kinds of enum values. I would say, it is also fine to return unexisting kinds in an internal API, since this fact can then be shared as a common knowledge in the team, so additional checks can be made on caller side if necessary. However, this is public API and I can say with a very high level of certainty that consumers of public API by default assume that if API returns an enum value, this is an actual existing value of that enum. So I would classify this as a bug regardless of the user scenario it was discovered in (and I provided the way I personally discovered it)

AlekseyTs commented:
> So I would classify this as a bug regardless of the user scenario it was discovered in (and I provided the way I personally discovered it)

I am afraid we still need an issue, especially as you classify the behavior as a bug.

DoctorKrolic commented:
> I am afraid we still need an issue, especially as you classify the behavior as a bug.

Filled https://github.com/dotnet/roslyn/issues/72300 and https://github.com/dotnet/roslyn/issues/72301. I'll put this PR into a draft state until both issues are discussed and approved as actual bugs. Personally I feel _very_ strongly about the first one being a bug and less strongly about the second one (but it is still not normal behavior IMO)

DoctorKrolic commented:
@AlekseyTs https://github.com/dotnet/roslyn/issues/72300 has been classified as a bug. I've reverted other changes in this PR, so it only fixes that issue. This is now ready for review

AlekseyTs commented on src/Compilers/CSharp/Portable/Syntax/SyntaxKindFacts.cs:
```diff
@@ -172,10 +176,26 @@ private static bool IsDebuggerSpecialPunctuation(SyntaxKind kind)
         public static IEnumerable<SyntaxKind> GetPunctuationKinds()
         {
-            for (int i = (int)SyntaxKind.TildeToken; i <= (int)SyntaxKind.GreaterThanGreaterThanGreaterThanEqualsToken; i++)
+            for (int i = (int)SyntaxKind.TildeToken; i <= (int)SyntaxKind.DotDotToken; i++)
```
>DotDotToken [](http://example.com/codeflow?start=74&length=11)

I think that in the SyntaxKind enum, around declarations of kinds that are hardcoded in implementation of this method, we should add comments describing what changes to the enum should lead to changes in this function as well.

AlekseyTs commented:
Done with review pass (commit 6)

AlekseyTs commented on src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs:
```diff
@@ -276,5 +278,20 @@ public void IsAttributeTargetSpecifier()
         }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/72300")]
+        public void TestAllKindsReturnedFromGetKindsMethodsExist()
```
> What way of testing do you think is better here?

I am leaning towards the first way, but the range should be defined differently by comparison to ```GetPunctuationKinds```. The test will have limited value if it is based on the same logic. Perhaps, based on the kind immediately preceding the first punctuation and the kind immediately following the last punctuation.

AlekseyTs commented on src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs:
```diff
@@ -276,5 +279,54 @@ public void IsAttributeTargetSpecifier()
+            Assert.Equal(0, method.GetParameters().Length);
+            Assert.Equal(typeof(IEnumerable<SyntaxKind>), method.ReturnType);
+
+            var kindsFromMethod = ((IEnumerable<SyntaxKind>)method.Invoke(null, null)).ToArray();
```
>ToArray [](http://example.com/codeflow?start=87&length=7)

A HashSet of integer values (not enum values) is probably be more efficient for the purpose of ```Contains``` checks.

AlekseyTs commented on src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs:
```diff
@@ -276,5 +279,54 @@ public void IsAttributeTargetSpecifier()
+            var kindsFromMethod = ((IEnumerable<SyntaxKind>)method.Invoke(null, null)).ToArray();
+
+            // Check that the value right before the expected lower bound isn't contained in the enumerable returned from the target method
+            Assert.DoesNotContain((SyntaxKind)((int)firstExpectedKind - 1), kindsFromMethod);
```
>Assert.DoesNotContain((SyntaxKind)((int)firstExpectedKind - 1), kindsFromMethod); [](http://example.com/codeflow?start=12&length=81)

It is not clear why we are checking only against two values outside of the range. What prevents us from iterating through the result and asserting all values are within the expected range?

AlekseyTs replied:
> It is not clear why we are checking only against two values outside of the range. What prevents us from iterating through the result and asserting all values are within the expected range?

I am not sure what has happened with this suggestion. It looks like instead of strengthening the check. I.e. making sure the returned kinds all are withing the expected range, the following checks were completely removed:
```
            // Check that the value right before the expected lower bound isn't contained in the enumerable returned from the target method
            Assert.DoesNotContain((SyntaxKind)((int)firstExpectedKind - 1), kindsFromMethod);
             // Check that the value right after the expected upper bound isn't contained in the enumerable returned from the target method
            Assert.DoesNotContain((SyntaxKind)((int)lastExpectedKind + 1), kindsFromMethod);
```

Could you, please, elaborate?

DoctorKrolic replied:
I initialy understood this review as "we can replace current logic with simple in-range check". After doing that you said that the tests are no longer representative since they won't fail if the method fails to return all expected kinds. At the point of partially reverting my changes I just forgot that this was a thing.

DoctorKrolic replied:
Done by removing expected members from the set and at the end asserting, that the set is empty. It does the same thing as "in range" check since if something isn't in range it is gonna be leftover in the set and thus the assertion would fail

AlekseyTs commented on src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs:
```diff
@@ -276,5 +279,54 @@ public void IsAttributeTargetSpecifier()
+        }
+
+        [Theory, WorkItem("https://github.com/dotnet/roslyn/issues/72300")]
+        [InlineData(nameof(SyntaxFacts.GetContextualKeywordKinds), SyntaxKind.YieldKeyword, SyntaxKind.FileKeyword)]
```
>SyntaxKind.FileKeyword [](http://example.com/codeflow?start=92&length=22)

It feels like the test will require less maintenance if the end range value is exclusive, i.e. ```ElifKeyword``` for this range.

AlekseyTs replied:
This thread was marked as resolved, but I didn't see any response in the code or otherwise. Reactivating.

AlekseyTs commented on src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs:
```diff
@@ -276,5 +279,54 @@ public void IsAttributeTargetSpecifier()
+
+        [Theory, WorkItem("https://github.com/dotnet/roslyn/issues/72300")]
+        [InlineData(nameof(SyntaxFacts.GetContextualKeywordKinds), SyntaxKind.YieldKeyword, SyntaxKind.FileKeyword)]
+        [InlineData(nameof(SyntaxFacts.GetPreprocessorKeywordKinds), SyntaxKind.ElifKeyword, SyntaxKind.RestoreKeyword)] // This case doesn't cover all possible preprocessor kinds
```
>// This case doesn't cover all possible preprocessor kinds [](http://example.com/codeflow?start=121&length=58)

Then, perhaps, it should have its own dedicated test that provides proper coverage?

AlekseyTs commented on src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs:
```diff
@@ -276,5 +278,58 @@ public void IsAttributeTargetSpecifier()
+
+            foreach (var kind in (IEnumerable<SyntaxKind>)method.Invoke(null, null))
+            {
+                Assert.InRange(kind, firstExpectedKind, lastExpectedKind);
```
>Assert.InRange(kind, firstExpectedKind, lastExpectedKind); [](http://example.com/codeflow?start=16&length=58)

I am not sure what is going on with this test in commit 9. Instead of making incremental improvements to the test, it appears that the test got changed in a way that it no longer serves its requested purpose - "We should try to come up with a test that would fail when ```GetPunctuationKinds``` fails to return all expected kinds." What am I missing.

AlekseyTs commented on src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs:
```diff
@@ -276,5 +278,58 @@ public void IsAttributeTargetSpecifier()
+            foreach (var kind in SyntaxFacts.GetPreprocessorKeywordKinds())
+            {
+                // Value must either be between ElifKeyword and RestoreKeyword or 4 exceptional keywords, which don't fall into that range
+                if (kind >= SyntaxKind.ElifKeyword && kind <= SyntaxKind.RestoreKeyword)
```
>SyntaxKind.RestoreKeyword [](http://example.com/codeflow?start=62&length=25)

Same suggestion as for the previous test in terms of future maintenance.

AlekseyTs commented on src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs:
```diff
@@ -276,5 +278,58 @@ public void IsAttributeTargetSpecifier()
+        }
+
+        [Fact]
+        public void TestGetPreprocessorKeywordKindsReturnsExpectedResults()
```
>TestGetPreprocessorKeywordKindsReturnsExpectedResults [](http://example.com/codeflow?start=20&length=53)

It looks like this test is also failing to fulfil the requested aspect - "a test that would fail when ```GetPreprocessorKeywordKinds``` fails to return all expected kinds"

AlekseyTs commented:
Done with review pass (commit 9)

AlekseyTs commented on src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs:
```diff
@@ -276,5 +279,63 @@ public void IsAttributeTargetSpecifier()
+        }
+
+        [Fact]
+        public void TestGetPreprocessorKeywordKindsReturnsExpectedResults()
```
>TestGetPreprocessorKeywordKindsReturnsExpectedResults [](http://example.com/codeflow?start=20&length=53)

Please strengthen the test by making sure the returned kinds all are withing the expected set.

DoctorKrolic replied:
Done by removing expected members from the set and at the end asserting, that the set is empty

AlekseyTs commented:
Done with review pass (commit 10)

DoctorKrolic commented on src/Compilers/CSharp/Portable/Syntax/SyntaxKindFacts.cs:
```diff
@@ -142,9 +145,10 @@ public static IEnumerable<SyntaxKind> GetPreprocessorKeywordKinds()
             yield return SyntaxKind.DefaultKeyword;
-            yield return SyntaxKind.HiddenKeyword;
+
             for (int i = (int)SyntaxKind.ElifKeyword; i <= (int)SyntaxKind.RestoreKeyword; i++)
```
The `SyntaxKind.ElifKeyword` - `SyntaxKind.RestoreKeyword` already contains `SyntaxKind.HiddenKeyword`, so I removed the duplicate. Found while working on tests

AlekseyTs reviewed (approved):
LGTM (commit 11), assuming CI is passing.

AlekseyTs commented:
@dotnet/roslyn-compiler For the second review.

jcouv commented on src/Compilers/CSharp/Portable/Syntax/SyntaxKindFacts.cs:
```diff
@@ -17,6 +19,7 @@ public static IEnumerable<SyntaxKind> GetReservedKeywordKinds()
         {
             for (int i = (int)SyntaxKind.BoolKeyword; i <= (int)SyntaxKind.ImplicitKeyword; i++)
             {
+                Debug.Assert(Enum.IsDefined(typeof(SyntaxKind), (SyntaxKind)i));
```
>(SyntaxKind)i [](http://example.com/codeflow?start=64&length=13)

nit: do we need the cast here?

DoctorKrolic replied:
As far as I remember, there is an exception if we don't case, let me double check

DoctorKrolic replied:
Yes, we do because the underlying type of `SyntaxKind` is `ushort`:

> System.ArgumentException : Enum underlying type and the object must be same type or object must be a String. Type passed in was 'System.Int32'; the enum underlying type was 'System.UInt16'.

jcouv replied:
Thanks

jcouv reviewed (approved):
LGTM Thanks (iteration 11)

jcouv commented:
/azp run

jcouv commented:
Merged/squashed. Thanks for the contribution!
